File: frontend/pages/stock_management.py
import streamlit as st
from utils.api_client import api_client
from utils.auth import check_auth, show_logout_button

# Check authentication first
check_auth()

# Show logout button in sidebar
show_logout_button()

# Helper function to round numbers to 1 decimal place
def round_number(value):
    """Round number to 1 decimal place, handling None values"""
    if value is None:
        return 0.0
    return round(float(value), 1)

st.set_page_config(
    page_title="Stock Management - Stocker",
    page_icon="📦",
    layout="wide"
)

st.title("📦 Stock Management")

# Items section
categories = api_client.get_categories()
items = api_client.get_items()
st.subheader("Items")
if items:
    st.subheader("Current Stock Levels")
    
    # Create tabs for different views
    tab1, tab2 = st.tabs(["📊 Stock Overview", "📝 Log Stock Counts"])
    
    with tab1:
        # Display all items in a table
        for category in categories:
            st.subheader(category.get('name', 'Unknown'))
            for item in items:
                if item.get('category_id') == category.get('id'):
                    quantity = round_number(item.get('quantity', 0))
                    unit = item.get('unit', '')
                    with st.expander(f"{item.get('name', 'Unknown')} - {quantity} {unit}"):
                        col1, col2, col3 = st.columns(3)
                        
                        with col1:
                            st.write(f"**Category:** {item.get('category_name', 'Unknown')}")
                        
                        with col2:
                            st.write(f"**Current Stock:** {quantity} {unit}")
                        
                        with col3:
                            threshold = round_number(item.get('restock_threshold', 0))
                            current = round_number(item.get('quantity', 0))
                            if current <= threshold:
                                st.error(f"⚠️ Low Stock (Threshold: {threshold})")
                            else:
                                st.success(f"✅ Good Stock (Threshold: {threshold})")
    
    with tab2:
        st.write("**Log Stock Counts:**")
        
        # Select item to log stock for
        item_names = [item.get('name', 'Unknown') for item in items]
        selected_item_name = st.selectbox("Select Item:", item_names)
        
        if selected_item_name:
            selected_item = next((item for item in items if item.get('name') == selected_item_name), None)
            
            if selected_item:
                current_quantity = round_number(selected_item.get('quantity', 0))
                unit = selected_item.get('unit', '')
                st.write(f"**Current Stock:** {current_quantity} {unit}")
                
                with st.form(f"log_stock_{selected_item.get('id')}"):
                    new_quantity = st.number_input(
                        "New Quantity", 
                        value=current_quantity, 
                        min_value=0.0,
                        step=0.1,
                        key=f"qty_{selected_item.get('id')}"
                    )
                    notes = st.text_input("Notes (optional)", key=f"notes_{selected_item.get('id')}")
                    staff_name = st.text_input("Staff Name (optional)", key=f"staff_name_{selected_item.get('id')}")
                    
                    if st.form_submit_button("Log Stock Count"):
                        result = api_client.log_stock(selected_item.get('id'), round_number(new_quantity), notes, staff_name)
                        if result:
                            st.success("Stock count logged successfully!")
                            st.rerun()
                        else:
                            st.error("Failed to log stock count. Please try again.")

else:
    st.info("No items found. Create some items first!")
    
    # Quick action to go back to main page
    if st.button("🏠 Back to Dashboard"):
        st.switch_page("main.py") 

File: frontend/src/components/StockManagement.js
import React, { useState, useEffect } from 'react';
import { FaBoxes, FaExclamationTriangle, FaCheckCircle } from 'react-icons/fa';
import apiClient from '../utils/apiClient';

const StockManagement = () => {
  const [items, setItems] = useState([]);
  const [categories, setCategories] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  const [activeTab, setActiveTab] = useState('overview');
  const [selectedItem, setSelectedItem] = useState(null);
  const [formData, setFormData] = useState({
    quantity: 0,
    notes: '',
    staffName: ''
  });
  const [successMessage, setSuccessMessage] = useState('');

  useEffect(() => {
    fetchData();
  }, []);

  const fetchData = async () => {
    try {
      setLoading(true);
      const [itemsData, categoriesData] = await Promise.all([
        apiClient.getItems(),
        apiClient.getCategories()
      ]);
      setItems(itemsData || []);
      setCategories(categoriesData || []);
    } catch (err) {
      setError('Failed to load data');
      console.error('Data error:', err);
    } finally {
      setLoading(false);
    }
  };

  const roundNumber = (value) => {
    if (value === null || value === undefined) return 0.0;
    return Math.round(parseFloat(value) * 10) / 10;
  };

  const handleLogStock = async (e) => {
    e.preventDefault();
    if (!selectedItem) return;

    try {
      await apiClient.logStock(
        selectedItem.id,
        roundNumber(formData.quantity),
        formData.notes || null,
        formData.staffName || null
      );
      setSuccessMessage('Stock count logged successfully!');
      setFormData({ quantity: 0, notes: '', staffName: '' });
      fetchData(); // Refresh data to get updated quantities
    } catch (err) {
      setError('Failed to log stock count. Please try again.');
    }
  };

  const handleItemSelect = (item) => {
    setSelectedItem(item);
    setFormData({
      quantity: roundNumber(item.quantity),
      notes: '',
      staffName: ''
    });
  };

  const clearMessages = () => {
    setError('');
    setSuccessMessage('');
  };

  if (loading) {
    return (
      <div style={{ textAlign: 'center', padding: '50px' }}>
        <div>Loading stock data...</div>
      </div>
    );
  }

  if (items.length === 0) {
    return (
      <div>
        <div style={{ marginBottom: '30px' }}>
          <h1 style={{ display: 'flex', alignItems: 'center', marginBottom: '10px' }}>
            <FaBoxes style={{ marginRight: '10px', color: '#007bff' }} />
            📦 Stock Management
          </h1>
        </div>
        <div className="alert alert-info">
          No items found. Create some items first!
        </div>
      </div>
    );
  }

  return (
    <div>
      <div style={{ marginBottom: '30px' }}>
        <h1 style={{ display: 'flex', alignItems: 'center', marginBottom: '10px' }}>
          <FaBoxes style={{ marginRight: '10px', color: '#007bff' }} />
          📦 Stock Management
        </h1>
      </div>

      {/* Tabs */}
      <div className="tabs">
        <button
          className={`tab ${activeTab === 'overview' ? 'active' : ''}`}
          onClick={() => { setActiveTab('overview'); clearMessages(); }}
        >
          📊 Stock Overview
        </button>
        <button
          className={`tab ${activeTab === 'log' ? 'active' : ''}`}
          onClick={() => { setActiveTab('log'); clearMessages(); }}
        >
          📝 Log Stock Counts
        </button>
      </div>

      {/* Messages */}
      {error && (
        <div className="alert alert-danger">
          {error}
        </div>
      )}
      {successMessage && (
        <div className="alert alert-success">
          {successMessage}
        </div>
      )}

      {/* Stock Overview Tab */}
      <div className={`tab-content ${activeTab === 'overview' ? 'active' : ''}`}>
        <div className="card">
          <h3>Current Stock Levels</h3>
          {categories.map((category) => {
            const categoryItems = items.filter(item => item.category_id === category.id);
            
            if (categoryItems.length === 0) return null;

            return (
              <div key={category.id} style={{ marginBottom: '30px' }}>
                <h4 style={{ 
                  borderBottom: '2px solid #007bff', 
                  paddingBottom: '5px',
                  marginBottom: '15px'
                }}>
                  {category.name}
                </h4>
                {categoryItems.map((item) => {
                  const quantity = roundNumber(item.quantity);
                  const unit = item.unit;
                  const threshold = roundNumber(item.restock_threshold);
                  const isLowStock = quantity <= threshold;

                  return (
                    <div key={item.id} className="expander">
                      <div className="expander-header">
                        <span>{item.name} - {quantity} {unit}</span>
                        {isLowStock ? (
                          <span style={{ color: '#dc3545' }}>
                            <FaExclamationTriangle style={{ marginRight: '5px' }} />
                            ⚠️ Low Stock
                          </span>
                        ) : (
                          <span style={{ color: '#28a745' }}>
                            <FaCheckCircle style={{ marginRight: '5px' }} />
                            ✅ Good Stock
                          </span>
                        )}
                      </div>
                      <div className="expander-content">
                        <div className="grid grid-3">
                          <div>
                            <strong>Category:</strong> {item.category_name}
                          </div>
                          <div>
                            <strong>Current Stock:</strong> {quantity} {unit}
                          </div>
                          <div>
                            <strong>Restock Threshold:</strong> {threshold} {unit}
                          </div>
                        </div>
                      </div>
                    </div>
                  );
                })}
              </div>
            );
          })}
        </div>
      </div>

      {/* Log Stock Counts Tab */}
      <div className={`tab-content ${activeTab === 'log' ? 'active' : ''}`}>
        <div className="card">
          <h3>Log Stock Counts</h3>
          
          <div className="form-group">
            <label>Select Item:</label>
            <select
              className="form-control"
              value={selectedItem?.id || ''}
              onChange={(e) => {
                const item = items.find(i => i.id === parseInt(e.target.value));
                if (item) {
                  handleItemSelect(item);
                }
              }}
            >
              <option value="">Select an item...</option>
              {items.map((item) => (
                <option key={item.id} value={item.id}>
                  {item.name} ({item.category_name})
                </option>
              ))}
            </select>
          </div>

          {selectedItem && (
            <div>
              <div style={{ 
                padding: '15px', 
                backgroundColor: '#f8f9fa', 
                borderRadius: '4px',
                marginBottom: '20px'
              }}>
                <h4>Current Stock Information</h4>
                <p><strong>Item:</strong> {selectedItem.name}</p>
                <p><strong>Category:</strong> {selectedItem.category_name}</p>
                <p><strong>Current Stock:</strong> {roundNumber(selectedItem.quantity)} {selectedItem.unit}</p>
                <p><strong>Restock Threshold:</strong> {roundNumber(selectedItem.restock_threshold)} {selectedItem.unit}</p>
              </div>

              <form onSubmit={handleLogStock}>
                <div className="form-group">
                  <label>New Quantity</label>
                  <input
                    type="number"
                    className="form-control"
                    value={formData.quantity}
                    onChange={(e) => setFormData({ ...formData, quantity: parseFloat(e.target.value) })}
                    min="0"
                    step="0.1"
                    required
                  />
                </div>

                <div className="form-group">
                  <label>Notes (optional)</label>
                  <input
                    type="text"
                    className="form-control"
                    value={formData.notes}
                    onChange={(e) => setFormData({ ...formData, notes: e.target.value })}
                    placeholder="Add any notes about this stock count"
                  />
                </div>

                <div className="form-group">
                  <label>Staff Name (optional)</label>
                  <input
                    type="text"
                    className="form-control"
                    value={formData.staffName}
                    onChange={(e) => setFormData({ ...formData, staffName: e.target.value })}
                    placeholder="Enter staff name"
                  />
                </div>

                <button type="submit" className="btn btn-primary">
                  Log Stock Count
                </button>
              </form>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default StockManagement; 